README.md
[Repository](https://github.com/pallet/haproxy-crate) &#xb7;
[Issues](https://github.com/pallet/haproxy-crate/issues) &#xb7;
[API docs](http://palletops.com/haproxy-crate/0.8/api) &#xb7;
[Annotated source](http://palletops.com/haproxy-crate/0.8/annotated/uberdoc.html) &#xb7;
[Release Notes](https://github.com/pallet/haproxy-crate/blob/develop/ReleaseNotes.md)

A [pallet](http://palletops.com/) crate to install and configure
 [haproxy](http://haproxy.1wt.eu/).

### Dependency Information

```clj
:dependencies [[com.palletops/haproxy-crate "0.8.0-alpha.3"]]
```

### Releases

<table>
<thead>
  <tr><th>Pallet</th><th>Crate Version</th><th>Repo</th><th>GroupId</th></tr>
</thead>
<tbody>
  <tr>
    <th>0.8.0-RC.4</th>
    <td>0.8.0-alpha.3</td>
    <td>clojars</td>
    <td>com.palletops</td>
    <td><a href='https://github.com/pallet/haproxy-crate/blob/0.8.0-alpha.3/ReleaseNotes.md'>Release Notes</a></td>
    <td><a href='https://github.com/pallet/haproxy-crate/blob/0.8.0-alpha.3/'>Source</a></td>
  </tr>
</tbody>
</table>

## Usage

The haproxy crate provides a `server-spec` function that returns a
server-spec. This server spec will install and run the haproxy server.
You pass a map of options to configure haproxy, as specified in
`settings`.

The `server-spec` provides an easy way of using the crate functions, and you can
use the following crate functions directly if you need to.

The `settings` function provides a plan function that should be called
in the `:settings` phase.  The function puts the configuration options
into the pallet session, where they can be found by the other crate
functions, or by other crates wanting to interact with haproxy.

The `:proxy-group` keyword can be used to provide a logical name the
haproxy instance, and defaults to pallet group name.

The haproxy configuration file is based on the `:config` key in the
settings map, and has four sub-keys, The `:listen` sub-key is used to
specific a map of applications to proxy, where the key is a keyword
naming the application, and the values are a map of configuration
options for the listen clause for that application.

For example, this would configure a `:my-app` application, listening
on port 80.

```clj
{:proxy-group :myproxy
 :config {:listen {:myapp {:server-address "0.0.0.0:80"}}}}
```

The `:frontend` and `:backend` keys can also be used to define an
application, e.g.

```clj
{:proxy-group :myproxy
 :config {:frontend {:myapp {:server-address "0.0.0.0:80"}}
          :backend {:myapp {}}}}
```

The `proxied-by` function should be called in the settings phase of
services that want to be proxied behind haproxy.  The `:proxy-group`
setting, which defaults to the group-name for the current target,
should match the `proxy-group-name` argument passed to `proxied-by`.

For example, this would add the current target as a server for the
`:my-app` application, contactable on the 8080 port.

```clj
(haproxy/proxied-by :proxy-group :myapp {:server-port 8080})
```

When the `:listen` key is used to set up the application, this will
result in server entries under in the listen configuration section.
When a `:backend` key is used to set up the application, this will
result in server entries under the backend configuration section.

The `install` function is responsible for actually installing haproxy.

## Live test on vmfest

For example, to run the live test on VMFest, using Ubuntu 13:

```sh
lein with-profile +vmfest pallet up --selectors ubuntu-13
lein with-profile +vmfest pallet down --selectors ubuntu-13
```

## License

Copyright (C) 2012, 2013 Hugo Duncan

Distributed under the Eclipse Public License.


ReleaseNotes.md
# Release notes

## 0.8.0-alpha.2

- Change ensure-service to use reload
  To ensure the service is up and running with the latest configuration,
  ensure-service now starts the service if stopped, and reloads the service
  if the configuration has changed.

## 0.8.0-alpha.2

- Fix use of frontend and backend configuration

## 0.8.0-alpha.1

- Update for pallet 0.8.0-RC.4

- Use settings phase

## haproxy-0.7.0

- Fix for clojure 1.3 compatibility

- Update for pallet 0.7.0

## haproxy-0.6.0

- Make sure the crate is idempotent. Allow options.

- Make Haproxy enable the 'epel' repo when installing.

## pallet-crates-0.5.0

## pallet-crates-0.4.4

- Update for repository management in separate namespaces

## pallet-crates-0.4.3

- Update for 0.5.0-SNAPSHOT
  Change pallet.resource.* to pallet.action.*. Change stevedore calls to
  script functions to use unquote and the pallet.script.lib namespace.
  Change request to session.  Change build-resources to build-actions.


## pallet-crates-0.4.2


## pallet-crates-0.4.1


## pallet-crates-0.4.0


## pallet-crates-0.4.0-beta-1

- Use epel for haproxy on CentOS

- fix haproxy for amzn-linux

- rationalise nagios and haproxy host naming

- Move node creation for tests in test-utils. Fix direct java style access to
  nodes from crates

- working node-list compute service

- Refactored to extract compute provider lib

- haproxy crate defaults to using epel 5-4 on centos

- Switch amzn-linux to use epel for haproxy

- Add centos 5.5 repo for haproxy on amzn-linux

- Updated the haproxy crate to use safe ids.

- Fixed issue in haproxy crate where it would break with EC2 since EC2 nodes
  have no name, only id.

- fix haproxy test

- only produce configuration for keys with content

- Updated to use template as a map, and for new Hardware in jclouds nodes

- Make the sequence of sections in haproxy config file deterministic

- enable haproxy in /etc/defaults

- add haproxy crate


test/pallet/crate/haproxy_test.clj
(ns pallet.crate.haproxy-test
  (:require
   [clojure.test :refer :all]
   [pallet.api :refer [group-spec plan-fn server-spec]]
   [pallet.compute :as compute]
   [pallet.build-actions :refer [build-actions build-session]]
   [pallet.actions :refer [remote-file]]
   [pallet.core.session :refer [with-session session]]
   [pallet.crate.etc-default :as etc-default]
   [pallet.crate.haproxy :as haproxy]
   [pallet.crate.network-service :refer [wait-for-port-listen]]
   [pallet.node :refer [id]]
   [pallet.test-utils :as test-utils]
   [pallet.utils :refer [base64-md5]]))

(deftest format-k-v-test
  (is (= "a b "
         (#'haproxy/format-kv :a "b" " ")))
  (is (= "a b\n"
         (#'haproxy/format-kv :a "b" \newline)))
  (is (= "a b\na c\n"
         (#'haproxy/format-kv :a ["b" "c"] \newline))))

(deftest config-server-test
  (is (= "tag 1.2.3.4:80 fall 3 check "
         (#'haproxy/config-server
          {:name :tag :ip "1.2.3.4" :server-port 80 :fall 3 :check true}))))

(deftest proxied-by-test
  (let [node (test-utils/make-node "tag" :public-ips ["1.2.3.4"])]
    (is (= {:host
            {"tag-1-2-3-4"
             {:haproxy {nil {:tag1 {:app1 {:name "tag", :ip "1.2.3.4"}}}}}}}
           (with-session
             (build-session {:server (assoc (group-spec :tag) :node node)})
             (haproxy/proxied-by :tag1 :app1 {})
             (select-keys (:plan-state (session)) [:host]))))))

(deftest server-listen-test
  (let [node (test-utils/make-node "tag" :public-ips ["1.2.3.4"])]
    (is (= {:app1 {:server ["tag 1.2.3.4 check "]
                   :balance "round-robin"
                   :server-address "0.0.0.0:80"}}
           (#'haproxy/server-listen
            :tag1
            {:app1 {:server-address "0.0.0.0:80" :balance "round-robin"}}
            {:app1 [{:name "tag", :ip "1.2.3.4" :check true}]})))
    (is (= {:app1 {:balance "round-robin"}}
           (#'haproxy/server-listen
            :tag1
            {:app1 {:balance "round-robin"}}
            {})))))

(deftest config-section-test
  (is (= (str
          "listen app1 0.0.0.0:80\n"
          "server tag 1.2.3.4 name tag \n"
          "balance round-robin\n\n")
         (#'pallet.crate.haproxy/config-section
          [:listen {:app1 {:server ["tag 1.2.3.4 name tag "]
                           :server-address "0.0.0.0:80"
                           :balance "round-robin"}}])))
  (is (= (str
          "listen app1 0.0.0.0:80\n"
          "server tag 1.2.3.4 name tag \n"
          "balance round-robin\n\n")
         (apply str
                (map
                 #'pallet.crate.haproxy/config-section
                 {:listen {:app1 {:server ["tag 1.2.3.4 name tag "]
                                  :server-address "0.0.0.0:80"
                                  :balance "round-robin"}}})))))

(deftest configure-test
  (is (=
       (first
        (build-actions
         {:server {:image {:os-family :ubuntu}
                   :group-name :tag
                   :node (test-utils/make-node
                          "tag" :public-ips ["1.2.3.4"])}
          :phase-context "configure"}
         (remote-file
          "/etc/haproxy/haproxy.cfg"
          :content "global\ngroup haproxy\nmaxconn 4096\nlog 127.0.0.1 local0\nlog 127.0.0.1 local1 notice\ndaemon\nuser haproxy\n\ndefaults\nclitimeout 50000\nretries 3\nmaxconn 2000\nlog global\nsrvtimeout 50000\ncontimeout 5000\nmode http\noption httplog\noption dontlognull\noption redispatch\n\nlisten app 0.0.0.0:80\nserver h1 1.2.3.4:80 weight 1 maxconn 50 check\nserver h2 1.2.3.5:80 weight 1 maxconn 50 check\n\n"
          :literal true
          :owner "haproxy"
          :group "haproxy"
          :flag-on-changed haproxy/config-changed-flag)
         (etc-default/write "haproxy" :ENABLED 1)))
       (first
        (build-actions
         {:server {:image {:os-family :ubuntu}
                   :group-name :tag
                   :node (test-utils/make-node
                          "tag" :public-ips ["1.2.3.4"])}}
         (haproxy/settings
          {:config
           {:listen {:app
                     {:server-address "0.0.0.0:80"
                      :server ["h1 1.2.3.4:80 weight 1 maxconn 50 check"
                               "h2 1.2.3.5:80 weight 1 maxconn 50 check"]}}
            :defaults {:mode "http"}}})
         (haproxy/configure {}))))))

(deftest invocation-test
  (let [node (test-utils/make-node "tag" :ip "1.2.3.4")]
    (is (build-actions
         {:server {:image {:os-family :ubuntu}
                   :group-name :tag :node node
                   :node-id (keyword (id node))}}
         (haproxy/settings
          {:config
           {:listen {:app
                     {:server-address "0.0.0.0:80"
                      :server ["h1 1.2.3.4:80 weight 1 maxconn 50 check"
                               "h2 1.2.3.5:80 weight 1 maxconn 50 check"]}}}})
         (haproxy/install {})
         (haproxy/proxied-by :tag :app {})
         (haproxy/configure {})))))

(def test-spec
  (server-spec
   :extends [(haproxy/server-spec
              {:config {:listen {:app {:server-address "0.0.0.0:8080"}}}})]
   :phases {:verify (plan-fn (wait-for-port-listen 8080))}
   :default-phases [:install :configure :ensure-service :verify]))


release.sh
#!/bin/bash

# release haproxy crate

if [[ $# -lt 3 ]]; then
  echo "usage: $(basename $0) previous-version new-version next-version" >&2
  exit 1
fi

previous_version=$1
version=$2
next_version=$3

echo ""
echo "Start release of $version, previous version is $previous_version"
echo ""
echo ""

lein do clean, test && \
git flow release start $version || exit 1

lein with-profile +release set-version ${version} :previous-version ${previous_version} \
  || { echo "set version failed" >2 ; exit 1; }

echo ""
echo ""
echo "Changes since $previous_version"
git --no-pager log --pretty=changelog $previous_version..
echo ""
echo ""

$EDITOR resources/pallet_crate/haproxy_crate/meta.edn
$EDITOR doc-src/USAGE.md
git add resources/pallet_crate/haproxy_crate/meta.edn doc-src/USAGE.md doc-src/FOOTER.md \
&& git commit -m "Updated metadata for $version" \
&& lein crate-doc \
|| exit 1

echo "Now edit project.clj, ReleaseNotes and README"

$EDITOR project.clj
$EDITOR ReleaseNotes.md
$EDITOR README.md

echo -n "commiting project.clj, release notes and readme.  enter to continue:" \
&& read x \
&& git add project.clj ReleaseNotes.md README.md \
&& git commit -m "Updated project.clj, release notes and readme for $version" \
&& echo -n "Peform release.  enter to continue:" && read x \
&& lein do clean, install, test, deploy clojars \
&& rm pom.xml \
&& git flow release finish $version \
&& echo "Now push to github. Don't forget the tags!" \
&& lein with-profile +doc doc \
&& lein with-profile +release set-version ${next_version} \
&& git add project.clj \
&& git commit -m "Updated version for next release cycle"


profiles.clj
{:dev
 {:dependencies [[com.palletops/pallet "0.8.0-RC.4" :classifier "tests"]
                 [com.palletops/crates "0.1.1"]
                 [ch.qos.logback/logback-classic "1.0.9"]]
  :plugins [[com.palletops/pallet-lein "0.8.0-alpha.1"]
            [com.palletops/lein-pallet-crate "0.1.0"]
            [lein-set-version "0.3.0"]
            [lein-resource "0.3.2"]
            [codox/codox.leiningen "0.6.4"]
            [lein-marginalia "0.7.1"]]
  :aliases {"live-test-up"
            ["pallet" "up"
             ;; "--phases" "install,configure,test"
             ;; "--selector" "live-test"
             ]
            "live-test-down" ["pallet" "down"]
            "live-test" ["do" "live-test-up," "live-test-down"]}
  :test-selectors {:default (complement :live-test)
                   :live-test :live-test
                   :all (constantly true)}}
 :doc {:dependencies [[com.palletops/pallet-codox "0.1.0"]]
       :plugins [[codox/codox.leiningen "0.6.4"]
                 [lein-marginalia "0.7.1"]]
       :codox {:writer codox-md.writer/write-docs
               :output-dir "doc/0.8/api"
               :src-dir-uri "https://github.com/pallet/haproxy-crate/blob/develop"
               :src-linenum-anchor-prefix "L"}
       :aliases {"marg" ["marg" "-d" "doc/0.8/annotated"]
                 "codox" ["doc"]
                 "doc" ["do" "codox," "marg"]}}
 :release
 {:set-version
  {:updates [{:path "README.md" :no-snapshot true}]}}
 :jclouds {:dependencies [[org.cloudhoist/pallet-jclouds "1.5.2"]
                          [org.jclouds/jclouds-allblobstore "1.5.5"]
                          [org.jclouds/jclouds-allcompute "1.5.5"]
                          [org.jclouds.driver/jclouds-slf4j "1.5.5"
                           :exclusions [org.slf4j/slf4j-api]]
                          [org.jclouds.driver/jclouds-sshj "1.5.5"]]}
 :vmfest {:dependencies [[com.palletops/pallet-vmfest "0.3.0-beta.2"]]}}


src/pallet/crate/haproxy.clj
(ns pallet.crate.haproxy
  "HA Proxy installation and configuration"
  (:require
   [clojure.set :refer [difference]]
   [clojure.string :as string]
   [clojure.tools.logging :refer [debugf warnf]]
   [pallet.actions :refer [remote-file]]
   [pallet.api :as api :refer [plan-fn]]
   [pallet.compute :as compute]
   [pallet.crate
    :refer [assoc-settings defplan get-node-settings get-settings
            group-name target-node target targets update-settings
            service-phases]]
   [pallet.crate-install :as crate-install]
   [pallet.crate.etc-default :as etc-default]
   [pallet.crate.initd :as initd]
   [pallet.crate.service :as service]
   [pallet.node :refer [hostname id primary-ip private-ip]]
   [pallet.utils :refer [apply-map base64-md5 deep-merge]]
   [pallet.version-dispatch
    :refer [defmethod-version-plan defmulti-version-plan]]))

;;; # Settings

(def config-changed-flag "HAPROXY-CONFIG-CHANGED")
(def facility :haproxy)

(def default-global
  {:log ["127.0.0.1 local0" "127.0.0.1 local1 notice"]
   :maxconn 4096
   :user "haproxy"
   :group "haproxy"
   :daemon true})

(def default-defaults
  {:log "global"
   :mode "http"
   :option ["httplog" "dontlognull" "redispatch"]
   :retries 3
   :maxconn 2000
   :contimeout 5000
   :clitimeout 50000
   :srvtimeout 50000})

(defmulti-version-plan default-settings [version])

;; By default, install from system packages
(defmethod-version-plan default-settings {:os :linux}
  [os os-version version]
  {:config {:global  default-global
            :defaults default-defaults}
   :conf-file "/etc/haproxy/haproxy.cfg"
   :user "haproxy"
   :group "haproxy"
   :service-name "haproxy"
   :supervisor :initd
   :install-strategy :packages
   :packages ["haproxy"]
   :package-options {:disable-service-start true}})

(defplan settings
  "Build the configuration settings by merging the user supplied ones
  with the OS-related install settings and the default config settings
  for HAProxy.

  Configuration is via the `:config` map.  `:global` and `:defaults`
  both take maps of keyword value pairs.  `:listen` takes a map where
  the keys are of the form \"name\" and contain a `:server-address`
  key with a string containing \"ip:port\", and other
  keyword/value. Servers for each listen section can be declared with
  the proxied-by function."
  [{:keys [instance-id version config] :as settings}]
  (let [settings (deep-merge
                  {:version (or version :latest)
                   :proxy-group (keyword (group-name))}
                  (default-settings :latest)
                  (dissoc settings :instance-id))]
    (debugf "haproxy settings %s" settings)
    (assoc-settings facility settings {:instance-id instance-id})))

;;; # Install

(defplan install
  [{:keys [instance-id]}]
  (crate-install/install facility instance-id))

;;; # Configure
(defmulti format-kv (fn format-kv-dispatch [k v & _] (class v)))

(defmethod format-kv :default
  [k v sep]
  (format "%s %s%s" (name k) v sep))

(defmethod format-kv clojure.lang.IPersistentVector
  [k v sep]
  (reduce (fn format-kv-vector [s value] (str s (format-kv k value sep))) "" v))

(defmethod format-kv clojure.lang.Sequential
  [k v sep]
  (reduce (fn format-kv-vector [s value] (str s (format-kv k value sep))) "" v))

(defmethod format-kv Boolean
  [k v sep]
  (when v (format "%s%s" (name k) sep)))

(defn- config-values
  "Format a map as key value pairs"
  [m]
  (apply str (for [[k v] m] (format-kv k v \newline))))

(defn- config-section
  [[key values]]
  (if (#{:frontend :backend :listen} key)
    (reduce
     #(str
       %1
       (format
        "%s %s %s\n%s%s\n"
        (name key) (name (first %2)) (or (:server-address (second %2)) "")
        (config-values (select-keys (second %2) [:acl]))
        (config-values (dissoc (second %2) :server-address :acl))))
     ""
     values)
    (format "%s\n%s\n" (name key) (config-values values))))

(defn- config-server
  "Format a server configuration line"
  [{:keys [ip server-port] :as config}]
  {:pre [(:name config) ip]}
  (format
   "%s %s%s %s"
   (name (:name config))
   ip
   (if server-port (str ":" server-port) "")
   (apply
    str
    (for [[k v] (dissoc config :server-port :ip :name)]
      (format-kv k v " ")))))

(defn- config-backend-server
  "Format a backend server configuration line"
  [{:keys [ip server-port] :as config}]
  {:pre [(:name config) ip]}
  (format
   "%s %s%s %s"
   (name (:name config))
   ip
   (if server-port (str ":" server-port) "")
   (apply
    str
    (for [[k v] (dissoc config :server-port :ip :name)]
      (format-kv k v " ")))))

(defn proxied-map
  "Build a map from app to sequence of proxied configuration for the given
  haproxy group."
  [haproxy-group]
  (reduce
   (fn [m target]
     (debugf "proxied-map %s %s"
             (get-node-settings (:node target) facility)
             haproxy-group)
     (if-let [proxied (-> (get-node-settings (:node target) facility)
                            haproxy-group)]
       (reduce
        (fn [m [app config]]
          (update-in m [app] (fnil conj []) config))
        m proxied)
       m))
   {} (targets)))

(defn server-listen
  "Build a map for the listen configuration of haproxy"
  [group-name listen proxied-config]
  {:pre [(keyword? group-name)
         (or (nil? listen) (map? listen))
         (map? proxied-config)]}
  (let [apps (map keyword (keys listen))
        listen (zipmap apps (vals listen))
        app-keys (keys proxied-config)
        unconfigured (difference (set app-keys) (set apps))
        no-nodes (difference (set app-keys) (set apps))]
    (doseq [app unconfigured]
      (warnf "Unconfigured proxy %s %s" group-name app))
    (doseq [app no-nodes]
      (warnf "Configured proxy %s %s with no servers" group-name app))
    (reduce
     (fn [listen [app app-servers]]
       (if (get listen app)
         (update-in listen [app :server]
                    (fn [servers]
                      (concat servers (map config-server app-servers))))
         listen))
     listen
     proxied-config)))

(defn server-backend
  "Build a map for the backend configuration of haproxy"
  [group-name backend proxied-config]
  {:pre [(keyword? group-name)
         (or (nil? backend) (map? backend))
         (map? proxied-config)]}
  (let [apps (map keyword (keys backend))
        backend (zipmap apps (vals backend))
        app-keys (keys proxied-config)
        unconfigured (difference (set app-keys) (set apps))
        no-nodes (difference (set app-keys) (set apps))]
    (doseq [app unconfigured]
      (warnf "Unconfigured proxy %s %s" group-name app))
    (doseq [app no-nodes]
      (warnf "Configured proxy %s %s with no servers" group-name app))
    (reduce
     (fn [backend [app app-servers]]
       (if (get backend app)
         (update-in backend [app :server]
                    (fn [servers]
                      (concat servers (map config-server app-servers))))
         backend))
     backend
     proxied-config)))

(defn config-file
  "Returns a string containing the configuration file contents."
  [group-name config]
  (let [config (-> config
                   (update-in
                    [:listen]
                    #(server-listen group-name % (proxied-map group-name)))
                   (update-in
                    [:backend]
                    #(server-backend
                      group-name % (proxied-map group-name))))]
    (debugf "proxied-map %s" (proxied-map group-name))
    (debugf "config %s" config)
    (->>
     (filter config [:global :defaults :listen :frontend :backend])
     (map (juxt identity config))
     (map config-section)
     string/join)))

(defplan configure
  "Configure HAProxy."
  [{:keys [instance-id] :as options}]
  (let [{:keys [conf-file config group proxy-group user]}
        (get-settings facility {:instance-id instance-id})]
    (remote-file
     conf-file
     :content (config-file proxy-group config)
     :literal true :owner user :group group
     :flag-on-changed config-changed-flag)
    (etc-default/write "haproxy" :ENABLED 1)))

(defn target-name [{:keys [node]}]
  ((some-fn hostname (comp base64-md5 id)) node))

(defn target-ip [{:keys [node]}]
  ((some-fn private-ip primary-ip) node))

(defn target-config-map
  [target config]
  (merge
   {:name (target-name target)
    :ip (target-ip target)}
   config))

(defn proxied-by
  "Declare that a node is proxied by the given haproxy server.  This
  should be called in a phase that runs before :configure (such as
  :settings).

      (proxied-by :haproxy :app1 {:check true})."
  [proxy-group-name app-group
   {:keys [server-port addr backup check cookie disabled fall id
           inter fastinter downinter maxqueue minconn port redir
           rise slowstart source track weight]
    :as config}
   & {:keys [instance-id] :as options}]
  {:pre [(keyword? app-group)]}
  (update-settings
   facility options assoc-in [(keyword proxy-group-name) app-group]
   (target-config-map (target) config)))

(defplan service
  [& {:keys [instance-id] :as options}]
  (let [{:keys [supervision-options] :as settings}
        (get-settings facility {:instance-id instance-id})]
    (service/service settings (merge supervision-options
                                     (dissoc options :instance-id)))))

(defplan ensure-service
  "Ensure the service is running and has read the latest configuration."
  [& {:keys [instance-id] :as options}]
  (service :instance-id instance-id
           :if-stopped true
           :action :start)
  (service :instance-id instance-id
           :if-flag config-changed-flag
           :action :reload))

(defn server-spec
  [{:keys [instance-id] :as settings}]
  (let [options (select-keys settings [:instance-id])]
    (api/server-spec
     :phases (merge
              {:settings (plan-fn
                          (pallet.crate.haproxy/settings settings))
               :install (plan-fn (install options))
               :configure (plan-fn (configure options))
               :ensure-service (plan-fn (apply-map ensure-service options))}
              (service-phases facility options service
                              :actions [:start :stop :reload :restart]))
     :default-phases [:install :configure :ensure-service]
     :roles (when-let [proxy-group (:proxy-group settings)]
              #{proxy-group}))))
